### DISCUSSION TITLE
Get all registered routes

### DISCUSSION THREAD

**Discussion by ilya-zlobintsev** [tokio-rs/axum#860] [Ideas]
Is it possible to get all the routes that are registered within a service/router? I want to be able to return the list of routes from a handler. 
`Router` doesn't seem to expose any public data which can be shared as state, so this would probably have to be implemented internally within axum.

**Comment by davidpdrsn** [comment]
Thats not currently possible. If/when we get OpenAPI support it'll be part of that. See https://github.com/tokio-rs/axum/issues/50

Until then you'd have to build your own thing that wraps a `Router` and also tracks which routes have been added.

**Reply by tanshion** [reply]
swagger is an example of this, and I also want to do permission control by getting all the routes...
Any good suggestions

**Reply by davidpdrsn** [reply]
axum doesn't have a way to get registered routes. You need to store that yourself somehow when building the router.

**Reply by ilya-zlobintsev** [reply]
For openapi you can use [aide](https://github.com/tamasfe/aide), which has its own router wrapper that saves the routes.
